src/auth/src/credentials/mds_credential.rs

```diff
@@ -0,0 +1,364 @@
+use reqwest::Client;
+use std::collections::HashMap;
+use std::env;
+
+const METADATA_FLAVOR_VALUE: &str = "Google";
+const METADATA_FLAVOR: &str = "metadata-flavor";
+
+lazy_static! {
+    // Use lazy_static to initialize the metadata URLs.
+    static ref _METADATA_ROOT: String = format!(
+        "http://{}/computeMetadata/v1/",
+        env::var("GCE_METADATA_HOST").unwrap_or_else(|_| {
+            env::var("GCE_METADATA_ROOT").unwrap_or_else(|_| "metadata.google.internal".to_string())
+        })
+    );
+}
+
+#[allow(dead_code)] // TODO(#442) - implementation in progress
+pub(crate) struct MDSCredential<T>
+where
+    T: TokenProvider,
+{
+    token_provider: T,
+}
+
+#[async_trait::async_trait]
+impl<T> Credential for MDSCredential<T>
+where
+    T: TokenProvider,
+{
+    async fn get_token(&mut self) -> Result<Token> {
+        self.token_provider.get_token().await
+    }
+
+    async fn get_headers(&mut self) -> Result<Vec<(HeaderName, HeaderValue)>> {
+        let token = self.get_token().await?;
+        let mut value = HeaderValue::from_str(&format!("{} {}", token.token_type, token.token))
+            .map_err(|e| CredentialError::new(false, e.into()))?;
+        value.set_sensitive(true);
+        Ok(vec![(AUTHORIZATION, value)])
+    }
+
+    async fn get_universe_domain(&mut self) -> Option<String> {
+        Some("googleapis.com".to_string())
+    }
+}
+
+#[allow(dead_code)] // TODO(#442) - implementation in progress
+#[derive(serde::Deserialize, serde::Serialize, PartialEq, Debug, Clone)]
+struct ServiceAccountInfo {
+    email: String,
+    scopes: Option<Vec<String>>,
+    aliases: Option<Vec<String>>,
+}
+#[allow(dead_code)] // TODO(#442) - implementation in progress
+struct MDSAccessTokenProvider {
+    token_endpoint: String,
+}
+
+#[allow(dead_code)]
+impl MDSAccessTokenProvider {
+    async fn get_service_account_info(
+        request: &Client,
+        metadata_service_endpoint: String,
+        email: Option<String>,
+    ) -> Result<ServiceAccountInfo> {
+        let email: String = email.unwrap_or("default".to_string());
+        let path: String = format!(
+            "{}/instance/service-accounts/{}/",
+            metadata_service_endpoint, email
+        );
+        let params = HashMap::from([("recursive", "true")]);
+
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            METADATA_FLAVOR,
+            HeaderValue::from_static(METADATA_FLAVOR_VALUE),
+        );
+
+        let url = reqwest::Url::parse_with_params(path.as_str(), params.iter())
+            .map_err(|e| CredentialError::new(false, e.into()))?;
+
+        let response = request
+            .get(url.clone())
+            .headers(headers)
+            .send()
+            .await
+            .map_err(|e| CredentialError::new(false, e.into()))?;
+
+        response
+            .json::<ServiceAccountInfo>()
+            .await
+            .map_err(|e| CredentialError::new(false, e.into()))
+    }
```

Comment:
All of this seems unrelated to the initial goal of getting a working MDS token.

Reply:
for fetching an mds token, first we need to fetch service account info specifically email and scopes, so therefore this call needs to happen first. As this was my first rust change request, so I wanted to keep PR requests small.

Reply:
I don't know what you mean.  This works:

curl -H"metadata-flavor:Google" http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/default/token

We do not need the email and/or scopes. Whatever are the defaults are good enough for us.

Reply:
Right now, I am following the python code here: https://github.com/googleapis/google-auth-library-python/blob/main/google/auth/compute_engine/credentials.py#L126 where I see first we fetch service account email and then we make the call to mds server with the email. The default call is also works, but will it be okay if we can review it as given I already have the pr out for it?

Reply:
Sounds good. I read your later comment, let me send another pull request for bare minimum for mds creds.